File: sdk-python/harpocrate/token.py
"""Parsing du token hrpv_* côté client — LOT_09 SDK.

Format : hrpv_{v}_{id_b32}_{exp_b36}_{perms_hex}_{auth_secret_b64url}_{dkey_b64url}_{hmac_b64url}

Le client :
  - Parse les champs pour en extraire wallet_id, decryption_key, permissions
  - Ne vérifie PAS le HMAC (la clé HMAC_KEY n'est pas côté client)
  - Vérifie l'expiration localement si exp != 0

Longueurs fixes (même convention que le backend) :
  - id_b32 : 26 chars (UUID 16 bytes → base32 lowercase sans padding)
  - auth_secret_b64 : 43 chars (32 bytes → base64url sans padding)
  - dkey_b64 : 43 chars (32 bytes → base64url sans padding)
  - hmac_b64 : 22 chars (16 bytes → base64url sans padding)
"""

from __future__ import annotations

import base64
import time
from dataclasses import dataclass
from uuid import UUID

from harpocrate.exceptions import InvalidTokenError, TokenExpiredError

_TOKEN_PREFIX = "hrpv"
_TOKEN_VERSION = "1"

# Longueurs fixes des champs encodés (identiques au backend)
_ID_B32_LEN = 26
_AUTH_SECRET_LEN = 43
_DKEY_LEN = 43
_HMAC_LEN = 22


@dataclass(frozen=True)
class ParsedToken:
    """Champs extraits d'un token hrpv_*."""

    version: str
    api_key_id: UUID
    exp: int  # timestamp Unix, 0 = pas d'expiration
    permissions: int  # bitmap
    auth_secret_b64: str  # raw field (43 chars)
    decryption_key: bytes  # 32 bytes décodés depuis dkey_b64
    dkey_b64: str  # raw field (43 chars)
    hmac_b64: str  # raw field — non vérifié côté client


def _b32_to_uuid(b32: str) -> UUID:
    """Décode un base32 lowercase 26 chars en UUID."""
    padded = (b32.upper() + "======")[:32]
    raw = base64.b32decode(padded)
    return UUID(bytes=raw)


def parse_token(token: str) -> ParsedToken:
    """Parse un token hrpv_* et retourne les champs décodés.

    Lève :
    - InvalidTokenError si le format est invalide
    - TokenExpiredError si le token est expiré

    N'effectue AUCUNE vérification HMAC (clé serveur uniquement).

    Stratégie de parsing positionnel : les 3 derniers champs ont des longueurs fixes
    pour gérer les '_' éventuels dans les encodages base64url.
    """
    if not isinstance(token, str) or not token.startswith(f"{_TOKEN_PREFIX}_"):
        raise InvalidTokenError("invalid_prefix", "Token must start with 'hrpv_'")

    # Longueur minimale
    min_len = (
        len(_TOKEN_PREFIX)
        + 1  # "hrpv_"
        + 1
        + 1  # version + "_"
        + _ID_B32_LEN
        + 1  # id + "_"
        + 1
        + 1  # exp (min 1 char) + "_"
        + 2
        + 1  # perms_hex + "_"
        + _AUTH_SECRET_LEN
        + 1  # auth + "_"
        + _DKEY_LEN
        + 1  # dkey + "_"
        + _HMAC_LEN  # hmac
    )
    if len(token) < min_len:
        raise InvalidTokenError("invalid_format", "Token is too short")

    # Extraction depuis la fin (champs à longueur fixe)
    hmac_b64 = token[-_HMAC_LEN:]
    if token[-(_HMAC_LEN + 1)] != "_":
        raise InvalidTokenError("invalid_format", "Malformed token structure")

    dkey_end = _HMAC_LEN + 1 + _DKEY_LEN
    dkey_b64 = token[-dkey_end : -(_HMAC_LEN + 1)]
    if token[-(dkey_end + 1)] != "_":
        raise InvalidTokenError("invalid_format", "Malformed token structure")

    auth_end = dkey_end + 1 + _AUTH_SECRET_LEN
    auth_secret_b64 = token[-auth_end : -(dkey_end + 1)]
    if token[-(auth_end + 1)] != "_":
        raise InvalidTokenError("invalid_format", "Malformed token structure")

    suffix_len = _AUTH_SECRET_LEN + 1 + _DKEY_LEN + 1 + _HMAC_LEN
    prefix_part = token[: -(suffix_len + 1)]

    early_parts = prefix_part.split("_")
    if len(early_parts) != 5:  # hrpv, v, id, exp, perms
        raise InvalidTokenError("invalid_format", "Malformed token structure (prefix)")

    prefix, version, id_b32, exp_b36, perms_hex = early_parts

    if prefix != _TOKEN_PREFIX:  # pragma: no cover
        raise InvalidTokenError("invalid_prefix", "Token must start with 'hrpv_'")

    if version != _TOKEN_VERSION:
        raise InvalidTokenError("unsupported_version", f"Unsupported token version: {version}")

    if len(id_b32) != _ID_B32_LEN:
        raise InvalidTokenError("invalid_id_encoding", "Invalid API key ID encoding")

    try:
        api_key_id = _b32_to_uuid(id_b32)
    except Exception as exc:
        raise InvalidTokenError("invalid_id_encoding", "Cannot decode API key ID") from exc

    try:
        exp = int(exp_b36, 36)
    except ValueError as exc:
        raise InvalidTokenError("invalid_exp_encoding", "Cannot decode expiration") from exc

    try:
        perms = int(perms_hex, 16)
    except ValueError as exc:
        raise InvalidTokenError("invalid_perms_encoding", "Cannot decode permissions") from exc

    if perms < 0 or perms > 0x3F:
        raise InvalidTokenError("invalid_perms_value", f"Permissions bitmap out of range: {perms}")

    try:
        # Ajoute le padding nécessaire pour base64url
        dkey_bytes = base64.urlsafe_b64decode(dkey_b64 + "==")
    except Exception as exc:
        raise InvalidTokenError("invalid_dkey_encoding", "Cannot decode decryption key") from exc

    if len(dkey_bytes) != 32:
        raise InvalidTokenError(
            "invalid_dkey_length",
            f"Decryption key must be 32 bytes, got {len(dkey_bytes)}",
        )

    # Vérification expiration (après décodage complet)
    if exp != 0 and exp < int(time.time()):
        raise TokenExpiredError("Token has expired")

    return ParsedToken(
        version=version,
        api_key_id=api_key_id,
        exp=exp,
        permissions=perms,
        auth_secret_b64=auth_secret_b64,
        decryption_key=dkey_bytes,
        dkey_b64=dkey_b64,
        hmac_b64=hmac_b64,
    )


def has_permission(parsed: ParsedToken, perm_bit: int) -> bool:
    """Vérifie si le token possède un bit de permission donné."""
    return bool(parsed.permissions & perm_bit)

File: sdk-typescript/src/token.ts
/**
 * Parsing du token `hrpv_v_id_exp_perms_auth_dkey_hmac` (format identique aux autres SDK).
 */
import { InvalidTokenError, TokenExpiredError } from './errors.js'

const PREFIX = 'hrpv'
const VERSION = '1'
const ID_B32_LEN = 26
const AUTH_SECRET_LEN = 43
const DKEY_LEN = 43
const HMAC_LEN = 22

export interface ParsedToken {
  version: string
  apiKeyId: string
  /** Timestamp Unix d'expiration. 0 = pas d'expiration. */
  exp: number
  /** Bitmap des permissions (max 0x3F). */
  permissions: number
  authSecretB64: string
  /** 32 bytes décodés depuis le champ dkey (clé AES-256 pour déchiffrer wallet_key). */
  decryptionKey: Uint8Array
  dkeyB64: string
  hmacB64: string
}

export function parseToken(token: string): ParsedToken {
  if (typeof token !== 'string' || !token.startsWith(`${PREFIX}_`)) {
    throw new InvalidTokenError('invalid_prefix', "Token must start with 'hrpv_'")
  }

  const suffixLen = AUTH_SECRET_LEN + 1 + DKEY_LEN + 1 + HMAC_LEN
  const minLen = PREFIX.length + 1 + 1 + 1 + ID_B32_LEN + 1 + 1 + 1 + 2 + 1 + suffixLen
  if (token.length < minLen) {
    throw new InvalidTokenError('invalid_format', 'Token is too short')
  }

  const total = token.length
  const hmacB64 = token.slice(total - HMAC_LEN)
  if (token[total - HMAC_LEN - 1] !== '_') {
    throw new InvalidTokenError('invalid_format', 'Malformed token structure')
  }
  const dkeyEnd = HMAC_LEN + 1 + DKEY_LEN
  const dkeyB64 = token.slice(total - dkeyEnd, total - HMAC_LEN - 1)
  if (token[total - dkeyEnd - 1] !== '_') {
    throw new InvalidTokenError('invalid_format', 'Malformed token structure')
  }
  const authEnd = dkeyEnd + 1 + AUTH_SECRET_LEN
  const authSecretB64 = token.slice(total - authEnd, total - dkeyEnd - 1)
  if (token[total - authEnd - 1] !== '_') {
    throw new InvalidTokenError('invalid_format', 'Malformed token structure')
  }

  const prefixPart = token.slice(0, total - suffixLen - 1)
  const parts = prefixPart.split('_')
  if (parts.length !== 5) {
    throw new InvalidTokenError('invalid_format', 'Malformed token structure (prefix)')
  }
  const [prefix, version, idB32, expB36, permsHex] = parts as [string, string, string, string, string]

  if (prefix !== PREFIX) {
    throw new InvalidTokenError('invalid_prefix', "Token must start with 'hrpv_'")
  }
  if (version !== VERSION) {
    throw new InvalidTokenError('unsupported_version', `Unsupported token version: ${version}`)
  }
  if (idB32.length !== ID_B32_LEN) {
    throw new InvalidTokenError('invalid_id_encoding', 'Invalid API key ID encoding')
  }

  const apiKeyId = decodeBase32Uuid(idB32)
  const exp = parseInt(expB36, 36)
  if (Number.isNaN(exp)) {
    throw new InvalidTokenError('invalid_exp_encoding', 'Cannot decode expiration')
  }
  const perms = parseInt(permsHex, 16)
  if (Number.isNaN(perms) || perms < 0 || perms > 0x3f) {
    throw new InvalidTokenError('invalid_perms_value', `Permissions out of range: ${permsHex}`)
  }

  const decryptionKey = base64UrlDecode(dkeyB64)
  if (decryptionKey.length !== 32) {
    throw new InvalidTokenError(
      'invalid_dkey_length',
      `Decryption key must be 32 bytes, got ${decryptionKey.length}`,
    )
  }

  if (exp !== 0 && exp < Math.floor(Date.now() / 1000)) {
    throw new TokenExpiredError()
  }

  return {
    version,
    apiKeyId,
    exp,
    permissions: perms,
    authSecretB64,
    decryptionKey,
    dkeyB64,
    hmacB64,
  }
}

// ─── Helpers ──────────────────────────────────────────────────────────────────

const B32_ALPHABET = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ234567'

function decodeBase32Uuid(b32: string): string {
  // RFC 4648 lowercase 26 chars → 16 bytes UUID
  const padded = (b32.toUpperCase() + '======').slice(0, 32)
  const bytes = base32Decode(padded)
  if (bytes.length !== 16) {
    throw new InvalidTokenError('invalid_id_encoding', 'Cannot decode API key ID')
  }
  // Format UUID standard 8-4-4-4-12
  const hex = Array.from(bytes, (b) => b.toString(16).padStart(2, '0')).join('')
  return `${hex.slice(0, 8)}-${hex.slice(8, 12)}-${hex.slice(12, 16)}-${hex.slice(16, 20)}-${hex.slice(20, 32)}`
}

function base32Decode(s: string): Uint8Array {
  const clean = s.replace(/=+$/, '')
  const out: number[] = []
  let buffer = 0
  let bitsLeft = 0
  for (const c of clean) {
    const idx = B32_ALPHABET.indexOf(c)
    if (idx < 0) {
      throw new InvalidTokenError('invalid_id_encoding', `Invalid base32 char: ${c}`)
    }
    buffer = (buffer << 5) | idx
    bitsLeft += 5
    if (bitsLeft >= 8) {
      bitsLeft -= 8
      out.push((buffer >> bitsLeft) & 0xff)
    }
  }
  return new Uint8Array(out)
}

function base64UrlDecode(s: string): Uint8Array {
  // Conversion base64url → base64 standard
  let padded = s.replace(/-/g, '+').replace(/_/g, '/')
  while (padded.length % 4 !== 0) padded += '='
  // atob disponible en Node 18+ et browser
  const binary = atob(padded)
  const bytes = new Uint8Array(binary.length)
  for (let i = 0; i < binary.length; i++) bytes[i] = binary.charCodeAt(i)
  return bytes
}
